### DISCUSSION TITLE
Initial thoughts on Flows

### DISCUSSION THREAD

**Discussion by david-williams-mpp** [postmanlabs/postman-flows#4] [Ideas]
Hello, this is looking like an excellent addition to Postman and I am very excited to start using it.

This feature is something I have been wanting for a very long time to solve a problem that I have which I will try to explain.

So in my example I have let's say we have 5 flows which all have 10 requests each. Of those 10, 8 are always the same. In order to progress from the first request to the second request, the first request must pass certain tests (otherwise there is no point to continue the flow). Also subsequent requests depend on some information from the previous request being passed in.

At this point we quite often end up writing tests in the requests such as:

``` js
pm.test('Valid response', () => {
    const responseJson = pm.response.json();
    pm.expect(responseJson.reference, 'Response should contain reference').to.exist.and.to.not.be.empty;
    pm.environment.set('reference', responseJson.reference);
    // Repeat last two lines for each variable required in the next request
}
```

This leads to quite a lot of repetition and also the exact same "tests" are duplicated for each of the 5 flows (which means if they need to be changed it is a lot more work).

In addition we then use Postman's "workflows" to control flow of requests when using the runner something like:

``` js
postman.setNextRequest(null);

pm.test('Test', () => {
    // Assertions
    postman.setNextRequest();
}
```

If the assertions fail the `setNextRequest()` will not be called so the flow will stop.

_Note: This limits us to one test per request otherwise if a test fails but a later test passes then it would allow the workflow to continue._

To assist with this I created a workflow lib which allows us to write something like:

``` js
workflow
    .start(pm, postman)
    .testHttpStatusCode(200)
    // Outputs reference into environment variables as 'reference'
    .testJsonProperty(json => json.reference, 'reference')
    .testJsonProperty(json => json.email, 'email')
    .run();
```

When this is run in the collection runner, if any of the tests fail the run will stop at the failed test. It will not run any further tests, nor will it run any further requests. The test output is easy to understand, and there is very little repetition.

_Note: There is a bunch of other functionality in the lib which makes it easy to build workflows which loop, skip etc. And also allows use of "iteration data" and provides the ability to run a specific iteration from the data using the collection runner._

However, due to Postman limitations:

a) We still have to duplicate each of the 8 common requests because workflows cannot move outside of their own folder (and having all requests in a single folder is a nightmare to manage).
b) It is a pain to import a custom lib (requires use of webpack and the eval trick in the pre-request script)

So the new "Flows" feature looks like it should basically solve a lot of these issues by allowing you to build up flows across different nested folders, and allowing the flow to be validated however at present it will not be possible to use this feature for my use case for the following reasons:

1. No way to feed data in from a file (e.g. iteration data like in the collection runner)
  _This is important because it allows us to test a variety of scenarios across the flow quickly without altering the flow each time_
    Suggestion: Add a new block type which can load from a file (bonus would be to allow such files to be part of the workspace so they can be shared) or add an option to "Create Variables" which allows it to load from file
    _A workaround for this could be to publish the iteration data somewhere so it can be fetched as part of a "Send Request" block, but that does require a process and somewhere to host said files._

    _One thing I thought about while writing this is to be able to add json files (nested under the flow) which can be selected when starting the flow. That way if someone wanted to use the same flow with a different config they could just duplicate the json and run the flow with the selected input._

2. Tests are not run which means you have limited control of the flow. For example you can currently terminate the flow by adding a validate block which checks status code, but the validation is much less flexible than the chai assertions in a standard test script.
    Suggestion: Add an option on "Send Request" blocks to run tests from the request and have pass/fail or success/error ports
    _and/or_
    Change the "Validate" block to a logical "If" block and add a new "Validate" block which allows you to run chai assertions and again has pass/fail ports

3. Inline with point 2, there is currently no output from the flow other than the usual console activity and the visualization of the flow itself. If tests are run, it would be good be able to see results similar to the collection runner. Of course it is possible to output to the terminal (and you still see requests in the console) but it would be nice if you could output request information without having to add terminal blocks to almost every block.
    Suggestion: Add "Terminal" as a dropdown in the popup "View information" say on a "Terminal" tab which is defaulted to "None" but allows you to select/create terminals. That way you can keep the flow diagram uncluttered and still get decent output. (bonus, would be nice to be able to configure the terminal output to display only relevant information or to be able to format it). For example, rather than showing the entire request, you might simply output "Reference: {{reference}}" (following earlier example").
    _One workaround is to shove a "Create Variables" block in before the terminal block wo reduce the data set, but that will obviously lead to even more clutter on the flow and still doesn't allow formatting_

4. Final point is around the "For Each" block. One of my other scenarios (which the lib mentioned above helps with) is when you want to build something which requires looping e.g. creating a cart and adding products. If you want to try a variety of carts with different contents it is currently possible to feed in iteration data and use `postman.setNextRequest(pm.info.requestName)` to loop. This requires writing code and is not easy for non-developers to follow. The "For Each" block would solve that nicely (I imagine looping over cart items and sending a request for each one, then moving on to the next request using the "finish" port). However there is again currently no convenient way to feed in your sample data to allow the looping.
    Suggestion: Similar to point 1. If we had the ability to feed in data this would be very powerful. One reason I listed this as it's own point is because it is important that the data can be fetched midway in the flow e.g. by using the "Create Variables" block to get something from iterationData. Currently the Create Variables only works with constants or output from the previous block. One way to handle this could be to add an additional option which lets you pull the data from any variables collection (including iterationData) to feed into the "For Each" block.

Other Issues:
- No undo functionality in the editor
- Start block cannot be moved and is not aligned to the grid so you can never get a straight line to the next block
- Input/output labels extend outside of the block so if you put blocks too close together you the input label of the next block overlaps the output of the previous block making it impossible to click the output to add another block
- Would be nice to be able to name/label blocks (and subsequently perhaps to be able to collapse large blocks e.g. with a lot of variables)
- Variables are not persisted across the flow. So a variable that is set by one request cannot be picked up by a subsequent request.
- Flow path is a little too subtle (might be good if you could have green/red (although that requires the ability to identify when a block failed, see points 2/3 above as they feed into this)
- Console seems to get two entries for each request in the flow.
- Not 100% convinced about starting the flow by clicking the start block. Can be clicked by accident when trying to move it :p
- The "On" port is a it confusing because it looks like a tick box and the hint is far away in the bottom right. So it took me a while to figure out what it does.
- No way to select multiple blocks which makes re-organising the flow (e.g. to insert a new block) quite difficult.
- Would be nice to pin the terminal outside of the flow so it doesn't cover up the flow while open

I realise this was quite a long read, but I think this feature has enormous potential and as I said I am very excited to make use of it in future so I am keen to help shape it if possible.

At the moment there is not functionality to export or share these flows publicly, however I assume this would be on the roadmap so as soon as it is I will create some examples to share.

Best regards

Dave

**Comment by saswatds** [comment]
Thanks a ton, @david-williams-mpp, for this valuable feedback. I will go ahead a create all relevant issues from your feedback. We will be able to track it better.
